src/state_machine.rs

```diff
@@ -355,21 +340,46 @@ impl StateMachine {
     #[cfg(feature = "use-mock-crust")]
     pub fn try_step(&mut self, outbox: &mut EventBox) -> Result<(), TryRecvError> {
         use itertools::Itertools;
+        use maidsafe_utilities::SeededRng;
+        use rand::Rng;
 
         if self.is_running {
             let mut events = Vec::new();
             while let Ok(category) = self.category_rx.try_recv() {
-                events.push(EventType::Category(category));
+                match category {
+                    MaidSafeEventCategory::Routing => {
+                        if let Ok(action) = self.action_rx.try_recv() {
+                            events.push(EventType::Action(Box::new(action)));
+                        } else {
+                            return Ok(self.apply_transition(Transition::Terminate, outbox));
+                        }
+                    }
+                    MaidSafeEventCategory::Crust => {
+                        if let Ok(crust_event) = self.crust_rx.try_recv() {
+                            events.push(EventType::CrustEvent(crust_event));
+                        } else {
+                            return Ok(self.apply_transition(Transition::Terminate, outbox));
+                        }
+                    }
+                }
             }
 
             let timed_out_events = self.state
-                .poll()
+                .get_timed_out_tokens()
                 .iter()
                 .map(|token| EventType::Action(Box::new(Action::Timeout(*token))))
                 .collect_vec();
 
+            if !events.is_empty() {
+                for event in timed_out_events {
+                    let pos = SeededRng::thread_rng().gen_range(0, events.len());
```

Comment:
This should probably be `events.len() + 1`, so they can also be inserted in last position. Then the distinction between empty and nonempty `events` would not be needed.

What I also find problematic is that this might now insert timer events in the wrong order. I think in production it is guaranteed that if you set a timer for 1 minute and another one for 2 minutes, you will handle the first one first. It's great if the tests pass this way, but I think it could cause surprising failures later, if we ever rely on the order of timer events.

Instead, you could do something like:
```rust
// Randomise where the timer events are inserted (zeroes).
let mut positions = iter::repeat(0)
    .take(timed_out_events.len())
    .chain(iter::repeat(1).take(events.len()))
    .collect_vec());
rng.shuffle(&mut positions);
self.events.extend(positions.map(|i| match i {
    0 => timed_out_events.pop(),
    1 => events.pop(),
}).rev());
```
(Or collect first, if that can't be `rev`ersed as is.)